Helm Chart Support (#310)
kube-green/kube-green: charts/README.md

### Install with Helm 

> 

Author: I put cert-manager helm install instructions. You can re-check?

> 

Reviewer: Would it be more suitable to define this chart as sub-chart dependency.

Helm allows you to do this via: https://v2.helm.sh/docs/developing_charts/#chart-dependencies

You can also make this conditional, so users that are bringing/using their own `cert-manager` won't install it twice.

```yaml
dependencies:
  - name: cert-manager
    version: 1.12.0
    repository: https://charts.jetstack.io
    condition: cert-manager.install
```

Looks like the cert-manager chart also allows specification of the namespace for use as a sub-chart:
https://github.com/cert-manager/cert-manager/blob/master/deploy/charts/cert-manager/values.yaml#L155

values.yaml
```yaml
...
cert-manager:
  install: true
  namespace: cert-manager
...
```

Happy to include this as a followup to these changes, to not overcomplicate things here :)

```bash
helm install \
--set image.repository=kubegreen/kube-green \

> --set image.tag=0.5.1 kube-green \

Author: Has been removed as requested

--namespace=kube-green \
--create-namespace=true \
./charts/kube-green
```

> ### Deploy Locally with Terraform

Author: Yeap! I tried to improve terraform details. You can review it again?

```hcl 
resource "helm_release" "kube_green" {
    namespace           = "kube-green"
    create_namespace    = true

    name                = "kube-green"
    chart               = "./charts/kube-green"

    set {
        name    = "image.repository"
        value   = "kubegreen/kube-green"
    }

    set {
        name    = "image.tag"
        value   = "0.5.1"
    }

    depends_on = [
        aws_eks_cluster.eks_cluster,
        kubernetes_config_map.aws-auth,
        aws_eks_fargate_profile.karpenter

>     ]

Author: Not necessarily, it's a copy / paste from my lab. Sorry 😆

# Install with Helm 

## Deploy cert-manager chart

```bash
helm repo add jetstack https://charts.jetstack.io

helm install \
  cert-manager jetstack/cert-manager \
  --namespace cert-manager \
  --create-namespace \
  --version v1.12.0 \
  --set installCRDs=true
```

> ##  Install kube-green chart 
> 

Reviewer 2: Since it could change with the time, to improve the maintainability we could set it as:

```suggestion
##  Install with kube-green chart 

To successfully install kube-green, in the cluster must be installed a cert-manager.
If it is not already installed installed, [check the cert-manager installation guide](https://cert-manager.io/docs/installation/).

To install kube-green using the helm-chart (inside the `kube-green` namespace), clone the kube-green repository and run this command:

```

> ## Deploy cert-manager chart
> 
> ```bash
> helm repo add jetstack https://charts.jetstack.io
> 
> helm install \
>   cert-manager jetstack/cert-manager \
>   --namespace cert-manager \
>   --create-namespace \
>   --version v1.12.0 \
>   --set installCRDs=true
> ```
> 
> ##  Install kube-green chart 
> 
> ```bash
> helm upgrade kube-green \
> --namespace=kube-green \
> --create-namespace=true \
> ./charts/kube-green --install 
> ```
> 

Reviewer 2: Since it could change with the time, to improve the maintainability we could set it as:

```suggestion
##  Install with kube-green chart 

To successfully install kube-green, in the cluster must be installed a cert-manager.
If it is not already installed installed, [check the cert-manager installation guide](https://cert-manager.io/docs/installation/).

To install kube-green using the helm-chart (inside the `kube-green` namespace), clone the kube-green repository and run this command:

```bash
helm upgrade kube-green \
--namespace=kube-green \
--create-namespace=true \
./charts/kube-green --install 
```

```

> 

Reviewer 2: ```suggestion
```

# Deploy Kube-Green Helm Chart with Terraform

> This example show how to use [Terraform Helm Chart Provider](https://developer.hashicorp.com/terraform/tutorials/kubernetes/helm-provider) to deploy `Kube-Green` on Kubernetes Clusters. 

Reviewer 2: ```suggestion
This example show how to use [Terraform Helm Chart Provider](https://developer.hashicorp.com/terraform/tutorials/kubernetes/helm-provider) to deploy `kube-green` on Kubernetes Clusters. 
```

## Prerequisites 
*   [Terraform Install](https://developer.hashicorp.com/terraform/tutorials/aws-get-started/install-cli)
*   [Helm Provider Credentials Setup](https://developer.hashicorp.com/terraform/tutorials/kubernetes/helm-provider#review-the-helm-configuration)

## Installation

> We need to install `cert-manager` as dependency before `kube-green` installation. To provision the both resources in same terraform run, you can declare helm release from cert_manager as dependency from kube-green helm helease using [depends_on](https://developer.hashicorp.com/terraform/language/meta-arguments/depends_on) meta-argument. 

Reviewer 2: ```suggestion
We need to install `cert-manager` as dependency before `kube-green` installation. To provision the both resources in same terraform run, you can declare helm release from cert_manager as dependency from kube-green helm release using [depends_on](https://developer.hashicorp.com/terraform/language/meta-arguments/depends_on) meta-argument. 
```

```hcl
resource "helm_release" "cert_manager" {
    namespace        = "cert-manager"
    create_namespace = true

    name       = "cert-manager"
    repository = "https://charts.jetstack.io"
    chart      = "cert-manager"
    version    = "v1.12.0"

    set {
        name  = "installCRDs"
        value = true
    }   

    set {
        name  = "webhook.timeoutSeconds"
        value = 10
    }
}
```

```hcl 
resource "helm_release" "kube_green" {
    namespace           = "kube-green"
    create_namespace    = true

    name                = "kube-green"
    chart               = "./charts/kube-green"

    set {
        name    = "image.repository"
        value   = "kubegreen/kube-green"
    }

    set {
        name    = "image.tag"
        value   = "0.5.1"
    }

    depends_on = [
        helm_release.cert_manager
    ]
}
```

> After configuration `helm_releases`, you can run terraform cli to provisioning kube-green installation properly. 

Reviewer 2: ```suggestion
After the configuration of the `helm_release`, you can run terraform cli to provisioning kube-green installation properly. 
```

# Install with Helm 

## Deploy cert-manager chart

```bash
helm repo add jetstack https://charts.jetstack.io

helm install \
  cert-manager jetstack/cert-manager \
  --namespace cert-manager \
  --create-namespace \
  --version v1.12.0 \
  --set installCRDs=true

> ```

Reviewer: Would it be more suitable to define this chart as sub-chart dependency.

Helm allows you to do this via: https://v2.helm.sh/docs/developing_charts/#chart-dependencies

You can also make this conditional, so users that are bringing/using their own `cert-manager` won't install it twice.

```yaml
dependencies:
  - name: cert-manager
    version: 1.12.0
    repository: https://charts.jetstack.io
    condition: cert-manager.install
```

Looks like the cert-manager chart also allows specification of the namespace for use as a sub-chart:
https://github.com/cert-manager/cert-manager/blob/master/deploy/charts/cert-manager/values.yaml#L155

values.yaml
```yaml
...
cert-manager:
  install: true
  namespace: cert-manager
...
```

Happy to include this as a followup to these changes, to not overcomplicate things here :)
